### PR TITLE
azure - common exception handling for actions

### PR DESCRIPTION
**Pull request by logachev**
- Refactor AzureBaseAction class to iterate over resources & handle CloudError exceptions
- Azure Actions should implement `_prepare_processing()` if needed and `_process_resource(resource)` functions
- Update Autotag action to get rid of code duplication
- Updated `execute_in_parallel` in parallel to disable multithreading

https://github.com/cloud-custodian/cloud-custodian/issues/3054

### CODE REVIEW

File: tools/c7n_azure/c7n_azure/actions.py [hunk 1]
```diff
@@ -60,18 +59,55 @@ def handle_exceptions(self, exceptions):
...
         )
 
+    def _process_resources(self, resources, event):
+        self._prepare_processing()
+
+        for r in resources:
+            try:
+                self._process_resource(r)
+            except CloudError as e:
+                TagHelper.log.error("Failed to process resource.\n"
```
Comment (erwelch) [inline tools/c7n_azure/c7n_azure/actions.py]: I know it's a WIP but just in case you didn't notice - not sure you mean to use the TageHelper log here?

Comment (logachev) [reply tools/c7n_azure/c7n_azure/actions.py]: this is why [wip] :)

Comment (erwelch) [reply tools/c7n_azure/c7n_azure/actions.py]: Alright I'll wait :D

**Comment by stefangordon** [comment]
I like this.

File: tools/c7n_azure/c7n_azure/utils.py [hunk 2]
```diff
@@ -104,31 +104,36 @@ class ThreadHelper:
...
-                        "Execution failed with error: %s" % f.exception())
-                    exceptions.append(f.exception())
-                    continue
-                else:
-                    result = f.result()
-                    if result:
-                        results.extend(result)
-
-            return results, list(set(exceptions))
+        if ThreadHelper.disable_multi_threading:
```
Comment (erwelch) [inline tools/c7n_azure/c7n_azure/utils.py]: I like this! Very clean.

File: tools/c7n_azure/c7n_azure/utils.py [hunk 3]
```diff
@@ -104,31 +104,36 @@ class ThreadHelper:
     disable_multi_threading = False
 
     @staticmethod
-    def execute_in_parallel(resources, execution_method, executor_factory, log,
+    def execute_in_parallel(resources, event, execution_method, executor_factory, log,
```
Comment (erwelch) [inline tools/c7n_azure/c7n_azure/utils.py]: Is event a required parameter?

Comment (erwelch) [reply tools/c7n_azure/c7n_azure/utils.py]: I suppose it gets passed through from process() in AzureActionBase which kicks all this off. Would we ever use this helper method other than in the AzureActionBase class where we might want to make the event optional here?

Comment (logachev) [reply tools/c7n_azure/c7n_azure/utils.py]: I wouldn't expect any other users for this method.. All Actions supposed to inherit from BaseAction and I don't know if parallel makes sense for filters. I assume, filters also can use this event.

File: tools/c7n_azure/c7n_azure/actions.py [hunk 4]
```diff
@@ -60,18 +60,55 @@ def handle_exceptions(self, exceptions):
...
+                               "Name: {1}.\n"
+                               "Error: {2}".format(r['type'], r['name'], e))
+
+    def _prepare_processing(self):
+        pass
+
+    @abc.abstractmethod
+    def _process_resource(self, resource):
+        raise NotImplementedError(
+            "Base action class does not implement behavior")
```
Comment (aluong) [inline tools/c7n_azure/c7n_azure/actions.py]: (minor) Base action class does not implement **this** behavior

File: tools/c7n_azure/c7n_azure/actions.py [hunk 5]
```diff
@@ -60,18 +60,55 @@ def handle_exceptions(self, exceptions):
...
+
+
+@six.add_metaclass(abc.ABCMeta)
+class AzureEventAction(EventAction, AzureBaseAction):
+
+    def _process_resources(self, resources, event):
+        self._prepare_processing()
```
Comment (aluong) [inline tools/c7n_azure/c7n_azure/actions.py]: Do you need to implement this? It should be available from AzureBaseAction already.

**Comment by aluong** [comment]
Loving where this is going.

**Pull request merged**